Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```diff
@@ -70,12 +72,73 @@
     HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
     parseMatchPatterns(origins, matchPatterns);
 
+    // If there is nothing to grant, return YES. This matches Chrome and Firefox.
+    if (!permissions.size() && !origins.size()) {
+        // FIXME: <https://webkit.org/b/250238> Invoke onAdded listeners.
+        completionHandler(true);
+        return;
+    }
+
     if (hasPermissions(permissions, matchPatterns)) {
         completionHandler(true);
         return;
     }
 
-    // FIXME: <https://webkit.org/b/250135> Call delegate method to prompt user for access.
+    // There shouldn't be any unsupported permissions, since they got reported as an error before this.
+    ASSERT(!permissions.isSubset(extension().supportedPermissions()));
+
+    bool requestedAllHostsPattern = false;
+    for (auto& matchPattern : matchPatterns) {
+        if (matchPattern->matchesAllHosts()) {
+            requestedAllHostsPattern = true;
+            break;
+        }
+    }
+
+    if (requestedAllHostsPattern && !m_requestedOptionalAccessToAllHosts)
+        m_requestedOptionalAccessToAllHosts = YES;
+
+    auto delegate = extensionController()->wrapper().delegate;
+
+    auto originsCompletionHandler = ^(NSSet<_WKWebExtensionMatchPattern *> *allowedOrigins) {
+        // User did not grant all requested origins.
+        if (allowedOrigins.count != origins.size()) {
+            completionHandler(false);
+            return;
+        }
+
+        // FIXME: Grant requested permissions
+
+        // FIXME: <https://webkit.org/b/250238> Add observer to _WKWebExtensionContextPermissionsWereGrantedNotification that invokes onAdded listeners.
+
+        completionHandler(true);
+    };
+
+    auto permissionCompletionHandler = ^(NSSet<_WKWebExtensionPermission> *allowedPermissions) {
+        // User did not grant all requested permissions.
```

Comment:
Figured we should return false if not all the permissions were granted. I can test in Chrome to see what the behavior is there.

Reply:
Unless the expected behavior is that the user either grants all the permissions or none at all?

Reply:
The client could grant some, or all.

Reply:
Tim and I decided to make the API require all requested permissions to be granted in order for it to return `true`.